File: packages/dataframe/rust/stats/distributions/poisson.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust Poisson functions...\n");

  const testCases = [
    { func: "dpois", distribution: "poisson", args: [3, 2, 0] },
    { func: "ppois", distribution: "poisson", args: [3, 2, 1, 0] },
    { func: "qpois", distribution: "poisson", args: [0.5, 2, 1, 0] },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rpois separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rpoisResult = await callR("rpois", "poisson", 1, 2);
    console.log(`rpois(2) = ${rpoisResult}`);
  } catch (error) {
    console.log(`rpois: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}

File: packages/dataframe/rust/stats/distributions/poisson.test.R
#!/usr/bin/env Rscript

# Poisson distribution test script
# Usage: Rscript poisson.test.R <function> <x_or_p> <lambda> [lower_tail] [log_p]

source("test-helpers.R")

# Define Poisson distribution functions
poisson_density <- function(x, lambda, log = FALSE) {
  dpois(x, lambda, log = log)
}

poisson_cumulative <- function(x, lambda, lower.tail = TRUE, log.p = FALSE) {
  ppois(x, lambda, lower.tail = lower.tail, log.p = log.p)
}

poisson_quantile <- function(p, lambda, lower.tail = TRUE, log.p = FALSE) {
  qpois(p, lambda, lower.tail = lower.tail, log.p = log.p)
}

poisson_random <- function(n, lambda) {
  rpois(n, lambda)
}

# Run the test
run_distribution_test(poisson_density, poisson_cumulative, poisson_quantile, poisson_random)
